### PR TITLE
Add delegated route efficiency report

### LINKED ISSUE #2764: workflow: add route-efficiency report for delegated agent work
## Goal / Problem

    Add a local route-efficiency report that measures whether delegated agent work is actually saving review effort.

    ## Archetype Metadata

    ```yaml
    archetype: analysis
    evidence_tier: analysis_only
    linked_policy:
  - docs/context/issue_1512_issue_archetypes.md
  - docs/context/artifact_evidence_vocabulary.md
  - docs/context/issue_691_benchmark_fallback_policy.md
    ```

    ## Context

    Source: attached rerun last-12-hour PR synthesis, current as of 2026-06-13.

    Delegation can appear useful while silently moving cost into review and recovery. The workflow now has manifests and snapshots that can be measured.

    ## Scope

    - Build a local report over routed-worker manifests and PR-loop snapshots without API calls.
- Summarize delegated attempts, artifact completeness, validation success, reroute count, parent-thread inspections avoided, and final outcome.
- Identify routes that produce incomplete artifacts or repeated failures.

    ## Non-goals

    - Do not require live GitHub or provider APIs.
- Do not include secrets, raw quota dumps, or full worker logs.
- Do not treat route success as task success.

    ## Definition of Done

    - [ ] Report emits JSON/Markdown from fixture or local manifest inputs.
- [ ] Route efficiency distinguishes complete artifacts, validation success, reroutes, and final accepted outcome.
- [ ] The report can guide future save-codex-tokens routing decisions.

    ## Validation / Testing

    - [ ] Run report tests on fixture manifests.
- [ ] Run a local dry-run report over available safe manifest samples if present.
- [ ] Inspect output for compactness and absence of secrets.

    ## Evidence Boundary

    Treat runtime, benchmark, metric, dissertation, and paper-facing claims as unproven until backed by durable, versioned evidence. Fallback, degraded, diagnostic-only, proxy-only, unavailable, stale, or missing-denominator rows must weaken or block claim wording.

    ## Project Metadata

    - Priority: Medium
    - Effort (h): 4-8
    - Reviewed: no

### PR DESCRIPTION
**Pull request by ll7**
## Summary
Adds a local route-efficiency report for routed-worker manifests so delegated worker attempts can be reviewed from compact artifacts instead of raw logs.

## Linked Issues
- Closes #2764

## Stack / Dependency
- Base dependency: none
- Required prior PRs: none
- Stack follow-up issues: none
- Safe to review independently: yes
- Review dependency reason, if any: NA

## What Changed
- Added `scripts/dev/route_efficiency_report.py` to summarize delegated attempts, complete artifact sets, reroutes, validation presence, and optional acceptance metadata from routed-worker manifests.
- Added focused unit tests plus a tracked representative fixture under `tests/fixtures/route_efficiency/`.
- Documented the compact report in `docs/dev_guide.md` beside the existing autopilot snapshot helpers.

## Why It Matters
- Added value: makes delegation quality measurable without pulling raw worker logs into the parent thread.
- Expected impact: supports cheaper Codex-controller workflows by exposing route completeness and reroute pressure from compact JSON/Markdown.
- Why this is worth merging now: recent goal-loop work is using more delegated CLIs, so this gives the workflow an immediate review-quality feedback surface.

## Research Result Guidance
- Target claim / hypothesis / blocker this should affect: NA - support helper for agent workflow review, not a research claim.
- Comparator or baseline, if applicable: NA
- Evidence tier: NA - support helper
- Result classification: NA
- Decision or stop rule, if applicable: NA
- Parent issue, claim map, registry, context note, or synthesis surface to update: #2764
- New research/benchmark/metric/paper-facing analysis tool, if any: NA - support helper; it summarizes delegation route artifacts and does not interpret research evidence.

## Falsification / Non-Transfer Check
- Did the mechanism activate? NA - support/tooling PR with no research mechanism claim.
- Did the intervention change command source, selected command, trajectory, or route progress? NA
- Did the scenario actually contain the targeted failure mode? NA
- Result route: NA
- Follow-up question or issue for weak, negative, or non-transfer results: NA

## Next Empirical Action
- Rerun needed: NA
- Extractor or analysis tool needed: no
- Artifact missing or unavailable: none
- Stop / revise / continue decision: stop; issue contract satisfied for local workflow reporting.
- Proposed child issue or existing follow-up: none

## Validation / Proof
- Commands run:
  - `uv run pytest tests/dev/test_route_efficiency_report.py tests/dev/test_routed_worker_manifest.py -q`
  - `uv run ruff check scripts/dev/route_efficiency_report.py tests/dev/test_route_efficiency_report.py`
  - `uv run ruff format --check scripts/dev/route_efficiency_report.py tests/dev/test_route_efficiency_report.py`
  - `git diff --check`
  - `uv run python scripts/dev/route_efficiency_report.py tests/fixtures/route_efficiency/routing_manifest_complete.json --format markdown`
  - `BASE_REF=origin/main scripts/dev/check_docs_proof_consistency_diff.sh`
  - `PYTEST_NUM_WORKERS=8 BASE_REF=origin/main scripts/dev/pr_ready_check.sh`
- Evidence that the change works here: fixture CLI emitted a compact Markdown report with 1/1 complete artifacts, validation present/success inferable, zero reroutes, and the route-evidence-only warning.
- Benchmarks or smoke tests, if applicable: full PR readiness passed with `6033 passed, 9 skipped, 10 warnings`; stamp `output/validation/pr_ready/issue-2764-route-efficiency-report.json`.

## Risks / Rollout
- Compatibility risks: low; new standalone CLI consumes existing `routed_worker_manifest.v1` fields and leaves existing tooling unchanged.
- Failure modes: validation success inference is heuristic when a manifest includes result text; manifests without result text still report validation presence without overclaiming success.
- Rollback or fallback plan: remove the new script/tests/docs pointer; no runtime behavior depends on it.

## Docs / Provenance
- Updated docs: `docs/dev_guide.md`
- Relevant design or provenance notes: active delegation ledger recorded partial OpenCode/Command Code routes and local validation.
- Any assumptions that need to be preserved: route success and artifact completeness remain route evidence only, not task acceptance.

## Downstream Propagation
- Parent issue updated (yes/no/NA): yes, via this PR closing #2764
- Claim map / benchmark report updated (yes/no/NA): NA
- Leaderboard / artifact catalog updated (yes/no/NA): NA
- Registry or config index updated (yes/no/NA): NA
- Context index / memory note updated (yes/no/NA): NA
- Follow-up issue opened for deferred propagation (yes/no/NA): NA
- Not applicable because: this is workflow support tooling, not benchmark/research evidence.

## Follow-Up Issues
- Deferred work: none
- Issues opened for follow-up: none

## Reviewer Notes
- Anything a reviewer should verify closely: route success is intentionally separated from task acceptance in both JSON and Markdown output.
- Any known limitations: `success_inferable` only increments when validation result text is present in the compact manifest; plain artifact-presence manifests report validation presence without claiming success.

## Summary by CodeRabbit

* **New Features**
  * Added a route efficiency reporting tool that analyzes worker routing manifests and generates reports showing delegation attempt counts, artifact completeness, validation results, and reroute tracking.

* **Documentation**
  * Updated developer guide with instructions for running and interpreting route efficiency reports.

* **Tests**
  * Added comprehensive test coverage for the reporting tool, including CLI behavior and edge cases.

### CODE REVIEW

File: scripts/dev/route_efficiency_report.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+
+def _extract_snapshot_outcome(
+    snapshot: dict[str, Any] | None,
+) -> tuple[bool | None, str | None]:
+    """Pull accepted and outcome from optional snapshot metadata."""
+    if not isinstance(snapshot, dict):
+        return None, None
+    accepted = bool(snapshot["accepted"]) if "accepted" in snapshot else None
+    note = str(snapshot["outcome"]) if "outcome" in snapshot else None
```
[the remark this answers is not in the document]
Comment (ll7) [reply scripts/dev/route_efficiency_report.py]: Fixed in 337989e7: snapshot accepted/outcome now check for None explicitly and preserve falsy values.

File: scripts/dev/route_efficiency_report.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+        self.total_attempts = 0
+        self.complete_count = 0
+        self.validation_present = 0
+        self.validation_success = 0
+        self.reroutes = 0
+        self.provider_incomplete: Counter[str] = Counter()
+        self.failure_class_incomplete: Counter[str] = Counter()
+
+    def process_attempt(self, idx: int, attempt: dict[str, Any]) -> None:
+        compact = attempt.get("compact_artifacts") or {}
```
[the remark this answers is not in the document]
Comment (ll7) [reply scripts/dev/route_efficiency_report.py]: Fixed in 337989e7: malformed attempts are handled defensively and counted as malformed incomplete attempts.

File: scripts/dev/route_efficiency_report.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    snapshot: dict[str, Any] | None = None,
+) -> dict[str, Any]:
+    """Build an efficiency report from one or more routed-worker manifests."""
+    acc = _Accumulator()
+
+    for manifest in manifests:
+        attempts = manifest.get("attempted_routes") or []
+        acc.total_attempts += len(attempts)
+        for idx, attempt in enumerate(attempts):
+            acc.process_attempt(idx, attempt)
```
[the remark this answers is not in the document]
Comment (ll7) [reply scripts/dev/route_efficiency_report.py:145]: Fixed in 59035eb8: manifest and attempted_routes type checks now match the suggested no-default pattern.

File: scripts/dev/route_efficiency_report.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+def main(argv: list[str] | None = None) -> int:
+    """CLI entry point."""
+    args = _parse_args(argv)
+    manifests: list[dict[str, Any]] = []
+    for path_str in args.manifests:
+        raw = json.loads(Path(path_str).read_text(encoding="utf-8"))
+        if isinstance(raw, list):
+            manifests.extend(raw)
+        else:
+            manifests.append(raw)
```
[the remark this answers is not in the document]
Comment (ll7) [reply scripts/dev/route_efficiency_report.py]: Fixed in 337989e7: CLI manifest loading now rejects non-object manifest payloads and list entries.
